lib/net.js

```diff
@@ -1166,6 +1248,79 @@ function lookupAndConnect(self, options) {
   debug('connect: dns options', dnsopts);
   self._host = host;
   const lookup = options.lookup || dns.lookup;
+
+  if (dnsopts.family !== 4 && dnsopts.family !== 6 && autoDetectFamily) {
+    debug('connect: autodetecting family via happy eyeballs');
+
+    dnsopts.all = true;
+
+    defaultTriggerAsyncIdScope(self[async_id_symbol], function() {
+      lookup(host, dnsopts, function emitLookup(err, addresses) {
+        // It's possible we were destroyed while looking this up.
+        // XXX it would be great if we could cancel the promise returned by
+        // the look up.
+        if (!self.connecting) {
+          return;
+        } else if (err) {
+          // net.createConnection() creates a net.Socket object and immediately
+          // calls net.Socket.connect() on it (that's us). There are no event
+          // listeners registered yet so defer the error event to the next tick.
+          process.nextTick(connectErrorNT, self, err);
+          return;
+        }
+
+        // This array will contain two elements at most, the first is a AAAA record, the second a A record
+        let ipv4Address;
+        let ipv6Address;
+
+        // Gather all the addresses we can use for happy eyeballs
+        for (let i = 0, l = addresses.length; i < l; i++) {
+          const address = addresses[i];
+          const { address: ip, family: addressType } = address;
+          self.emit('lookup', err, ip, addressType, host);
+
+          if (isIP(ip)) {
+            if (addressType === 6 && !ipv6Address) {
+              ipv6Address = address;
+            } else if (addressType === 4 && !ipv4Address) {
+              ipv4Address = address;
+            }
+          }
+
+          if (ipv6Address && ipv4Address) {
+            break;
+          }
+        }
+
+        // When no AAAA or A records are available, fail on the first one
+        if (!ipv6Address && !ipv4Address) {
+          const { address: firstIp, family: firstAddressType } = addresses[0];
+
+          if (!isIP(firstIp)) {
+            err = new ERR_INVALID_IP_ADDRESS(firstIp);
+            process.nextTick(connectErrorNT, self, err);
+          } else if (firstAddressType !== 4 && firstAddressType !== 6) {
+            err = new ERR_INVALID_ADDRESS_FAMILY(firstAddressType,
+                                                 options.host,
+                                                 options.port);
+            process.nextTick(connectErrorNT, self, err);
+          }
+
+          return;
+        }
+
+        self._unrefTimer();
+        defaultTriggerAsyncIdScope(
+          self[async_id_symbol],
+          internalConnectMultiple,
+          self, [ipv6Address, ipv4Address], port, localPort
+        );
+      });
+    });
```

Comment:
The `lookupAndConnect` function is already incredibly long. Could you extract all the logic so that it works as:

```js
if (dnsopts.family !== 4 && dnsopts.family !== 6 && autoDetectFamily) {
  autoDetectFamily(..)
}
```